File: gateway/src/quota/store.ts
import type postgres from "postgres";
import type { GatewayConfig } from "../config.js";
import { SESSION_CHANNEL, TURN_SPEAKER } from "../schema.js";
import { firstQuotaHit, quotaShouldWarn } from "./decision.js";
import {
  type QuotaLimits,
  envQuotaLimits,
  resolveQuotaLimits,
} from "./settings.js";

type Sql = ReturnType<typeof postgres>;

export type QuotaUsage = {
  turnsUsed: number;
  minutesUsed: number;
  resetAt: Date;
};

export type QuotaRefusal = {
  error: string;
  code: string;
  reset_at: string;
};

export async function loadQuotaSettings(sql: Sql): Promise<QuotaLimits | null> {
  const rows = await sql<
    {
      turns_per_day: number;
      voice_minutes_per_day: number;
      timezone: string;
      warn_ratio: number;
    }[]
  >`
    SELECT turns_per_day, voice_minutes_per_day, timezone, warn_ratio
    FROM quota_settings
    WHERE singleton
    LIMIT 1
  `;
  const row = rows[0];
  if (!row) {
    return null;
  }
  return {
    turnsPerDay: Number(row.turns_per_day),
    voiceMinutesPerDay: Number(row.voice_minutes_per_day),
    timezone: row.timezone,
    warnRatio: Number(row.warn_ratio),
  };
}

export async function saveQuotaSettings(
  sql: Sql,
  limits: QuotaLimits,
  actorId: string,
): Promise<QuotaLimits> {
  await sql`
    INSERT INTO quota_settings (
      singleton,
      turns_per_day,
      voice_minutes_per_day,
      timezone,
      warn_ratio,
      updated_at,
      updated_by
    ) VALUES (
      true,
      ${limits.turnsPerDay},
      ${limits.voiceMinutesPerDay},
      ${limits.timezone},
      ${limits.warnRatio},
      now(),
      ${actorId}::uuid
    )
    ON CONFLICT (singleton) DO UPDATE SET
      turns_per_day = EXCLUDED.turns_per_day,
      voice_minutes_per_day = EXCLUDED.voice_minutes_per_day,
      timezone = EXCLUDED.timezone,
      warn_ratio = EXCLUDED.warn_ratio,
      updated_at = now(),
      updated_by = EXCLUDED.updated_by
  `;
  return limits;
}

export async function resolveLiveQuotaLimits(
  sql: Sql,
  config: GatewayConfig,
): Promise<QuotaLimits> {
  return resolveQuotaLimits(
    await loadQuotaSettings(sql),
    envQuotaLimits(config),
  );
}

export async function loadQuotaUsage(
  sql: Sql,
  userId: string,
  limits: QuotaLimits,
): Promise<QuotaUsage> {
  const tz = limits.timezone;
  const window = await sql<{ window_start: Date; reset_at: Date }[]>`
    SELECT
      (date_trunc('day', timezone(${tz}, now())) AT TIME ZONE ${tz}) AS window_start,
      (
        date_trunc('day', timezone(${tz}, now())) AT TIME ZONE ${tz}
        + interval '1 day'
      ) AS reset_at
  `;
  const bounds = window[0];
  if (!bounds) {
    throw new Error("quota window query returned no row");
  }
  const start = bounds.window_start;
  const resetAt = bounds.reset_at;
  const turns = await sql<{ turns_used: number }[]>`
    SELECT count(*)::int AS turns_used
    FROM turns t
    INNER JOIN sessions s ON s.id = t.session_id
    WHERE s.user_id = ${userId}
      AND t.speaker = ${TURN_SPEAKER.USER}
      AND t.created_at >= ${start}
      AND t.created_at < ${resetAt}
  `;
  const minutes = await sql<{ minutes_used: number }[]>`
    SELECT COALESCE(
      SUM(
        GREATEST(
          0,
          EXTRACT(
            EPOCH FROM (
              LEAST(COALESCE(s.ended_at, now()), ${resetAt})
              - GREATEST(s.started_at, ${start})
            )
          )
        )
      ),
      0
    ) / 60.0 AS minutes_used
    FROM sessions s
    WHERE s.user_id = ${userId}
      AND s.channel = ${SESSION_CHANNEL.VOICE}
      AND s.started_at < ${resetAt}
      AND COALESCE(s.ended_at, now()) > ${start}
  `;
  return {
    turnsUsed: Number(turns[0]?.turns_used ?? 0),
    minutesUsed: Number(minutes[0]?.minutes_used ?? 0),
    resetAt,
  };
}

export function quotaRefusal(
  config: GatewayConfig,
  usage: QuotaUsage,
  limits: QuotaLimits,
): QuotaRefusal | null {
  const hit = firstQuotaHit({
    turnsUsed: usage.turnsUsed,
    turnsLimit: limits.turnsPerDay,
    turnsKind: config.QUOTA_KIND_TURNS,
    minutesUsed: usage.minutesUsed,
    minutesLimit: limits.voiceMinutesPerDay,
    minutesKind: config.QUOTA_KIND_MINUTES,
  });
  if (!hit) {
    return null;
  }
  const resetAt = usage.resetAt.toISOString();
  if (hit === config.QUOTA_KIND_TURNS) {
    return {
      error: config.QUOTA_ERROR_TURNS,
      code: config.QUOTA_CODE_TURNS,
      reset_at: resetAt,
    };
  }
  return {
    error: config.QUOTA_ERROR_MINUTES,
    code: config.QUOTA_CODE_MINUTES,
    reset_at: resetAt,
  };
}

export function quotaWarnKinds(
  config: GatewayConfig,
  usage: QuotaUsage,
  limits: QuotaLimits,
): string[] {
  const kinds: string[] = [];
  if (quotaShouldWarn(usage.turnsUsed, limits.turnsPerDay, limits.warnRatio)) {
    kinds.push(config.QUOTA_KIND_TURNS);
  }
  if (
    quotaShouldWarn(
      usage.minutesUsed,
      limits.voiceMinutesPerDay,
      limits.warnRatio,
    )
  ) {
    kinds.push(config.QUOTA_KIND_MINUTES);
  }
  return kinds;
}

File: worker/src/worker/quota/store.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any
from uuid import UUID

import psycopg
import structlog

from worker.config import WorkerSettings
from worker.quota.decision import first_quota_hit, quota_should_warn
from worker.quota.settings import QuotaLimits, env_quota_limits, resolve_quota_limits
from worker.schema import SESSION_CHANNEL_VOICE, TURN_SPEAKER_USER

log = structlog.get_logger(__name__)


@dataclass(frozen=True)
class QuotaUsage:
    turns_used: int
    minutes_used: float
    reset_at: datetime


def _window(conn: psycopg.Connection, timezone_name: str) -> tuple[datetime, datetime]:
    with conn.cursor() as cur:
        cur.execute(
            """
            SELECT
              (date_trunc('day', timezone(%s, now())) AT TIME ZONE %s) AS window_start,
              (
                date_trunc('day', timezone(%s, now())) AT TIME ZONE %s
                + interval '1 day'
              ) AS reset_at
            """,
            (timezone_name, timezone_name, timezone_name, timezone_name),
        )
        row = cur.fetchone()
    if row is None:
        raise RuntimeError("quota window query returned no row")
    return row["window_start"], row["reset_at"]


def load_quota_settings(conn: psycopg.Connection) -> QuotaLimits | None:
    with conn.cursor() as cur:
        cur.execute(
            """
            SELECT turns_per_day, voice_minutes_per_day, timezone, warn_ratio
            FROM quota_settings
            WHERE singleton
            LIMIT 1
            """
        )
        row = cur.fetchone()
    if row is None:
        return None
    return QuotaLimits(
        turns_per_day=int(row["turns_per_day"]),
        voice_minutes_per_day=float(row["voice_minutes_per_day"]),
        timezone=str(row["timezone"]),
        warn_ratio=float(row["warn_ratio"]),
    )


def resolve_live_quota_limits(
    conn: psycopg.Connection,
    settings: WorkerSettings,
) -> QuotaLimits:
    return resolve_quota_limits(load_quota_settings(conn), env_quota_limits(settings))


def load_quota_usage(
    conn: psycopg.Connection,
    user_id: UUID,
    limits: QuotaLimits,
) -> QuotaUsage:
    start, reset_at = _window(conn, limits.timezone)
    with conn.cursor() as cur:
        cur.execute(
            """
            SELECT count(*)::int AS turns_used
            FROM turns t
            INNER JOIN sessions s ON s.id = t.session_id
            WHERE s.user_id = %s
              AND t.speaker = %s
              AND t.created_at >= %s
              AND t.created_at < %s
            """,
            (str(user_id), TURN_SPEAKER_USER, start, reset_at),
        )
        turns_row = cur.fetchone()
        cur.execute(
            """
            SELECT COALESCE(
              SUM(
                GREATEST(
                  0,
                  EXTRACT(
                    EPOCH FROM (
                      LEAST(COALESCE(s.ended_at, now()), %s)
                      - GREATEST(s.started_at, %s)
                    )
                  )
                )
              ),
              0
            ) / 60.0 AS minutes_used
            FROM sessions s
            WHERE s.user_id = %s
              AND s.channel = %s
              AND s.started_at < %s
              AND COALESCE(s.ended_at, now()) > %s
            """,
            (reset_at, start, str(user_id), SESSION_CHANNEL_VOICE, reset_at, start),
        )
        minutes_row = cur.fetchone()
    turns_used = int(turns_row["turns_used"]) if turns_row else 0
    minutes_used = float(minutes_row["minutes_used"]) if minutes_row else 0.0
    return QuotaUsage(
        turns_used=turns_used,
        minutes_used=minutes_used,
        reset_at=reset_at,
    )


def log_quota_warn(
    settings: WorkerSettings,
    user_id: UUID,
    usage: QuotaUsage,
    limits: QuotaLimits,
) -> None:
    if quota_should_warn(usage.turns_used, limits.turns_per_day, limits.warn_ratio):
        log.warning(
            settings.quota_log_warn,
            user_id=str(user_id),
            kind=settings.quota_kind_turns,
            used=usage.turns_used,
            limit=limits.turns_per_day,
            ratio=limits.warn_ratio,
        )
    if quota_should_warn(
        usage.minutes_used,
        limits.voice_minutes_per_day,
        limits.warn_ratio,
    ):
        log.warning(
            settings.quota_log_warn,
            user_id=str(user_id),
            kind=settings.quota_kind_minutes,
            used=usage.minutes_used,
            limit=limits.voice_minutes_per_day,
            ratio=limits.warn_ratio,
        )


def refuse_quota(
    settings: WorkerSettings,
    usage: QuotaUsage,
    limits: QuotaLimits,
) -> dict[str, Any] | None:
    hit = first_quota_hit(
        turns_used=usage.turns_used,
        turns_limit=limits.turns_per_day,
        turns_kind=settings.quota_kind_turns,
        minutes_used=usage.minutes_used,
        minutes_limit=limits.voice_minutes_per_day,
        minutes_kind=settings.quota_kind_minutes,
    )
    if hit is None:
        return None
    if hit == settings.quota_kind_turns:
        return {
            "error": settings.quota_error_turns,
            "code": settings.quota_code_turns,
            "kind": settings.quota_kind_turns,
            "used": usage.turns_used,
            "limit": limits.turns_per_day,
        }
    return {
        "error": settings.quota_error_minutes,
        "code": settings.quota_code_minutes,
        "kind": settings.quota_kind_minutes,
        "used": usage.minutes_used,
        "limit": limits.voice_minutes_per_day,
    }
